src/geoFire/geoFireUtils.ts

```diff
@@ -0,0 +1,477 @@
+  }
+  geohash = geohash.substring(0, precision);
+  const base = geohash.substring(0, geohash.length - 1);
+  const lastValue = g_BASE32.indexOf(geohash.charAt(geohash.length - 1));
+  const significantBits = bits - (base.length * g_BITS_PER_CHAR);
+  const unusedBits = (g_BITS_PER_CHAR - significantBits);
+  // delete unused bits
+  const startValue = (lastValue >> unusedBits) << unusedBits;
+  const endValue = startValue + (1 << unusedBits);
+  if (endValue > 31) {
+    return [base + g_BASE32[startValue], base + '~'];
+  } else {
+    return [base + g_BASE32[startValue], base + g_BASE32[endValue]];
+  }
+};
+
+/**
+ * Calculates a set of queries to fully contain a given circle. A query is a [start, end] pair
+ * where any geohash is guaranteed to be lexiographically larger then start and smaller than end.
+ *
+ * @param center The center given as [latitude, longitude] pair.
+ * @param radius The radius of the circle.
+ * @return An array of geohashes containing a [start, end] pair.
+ */
+export function geohashQueries(center: number[], radius: number): string[][] {
+  validateLocation(center);
+  const queryBits = Math.max(1, boundingBoxBits(center, radius));
+  const geohashPrecision = Math.ceil(queryBits / g_BITS_PER_CHAR);
+  const coordinates = boundingBoxCoordinates(center, radius);
+  const queries = coordinates.map(function (coordinate) {
+    return geohashQuery(encodeGeohash(coordinate, geohashPrecision), queryBits);
+  });
+  // remove duplicates
+  return queries.filter(function (query, index) {
+    return !queries.some(function (other, otherIndex) {
+      return index > otherIndex && query[0] === other[0] && query[1] === other[1];
+    });
+  });
+};
+
+/**
+ * Encodes a location and geohash as a GeoFire object.
+ *
+ * @param location The location as [latitude, longitude] pair.
+ * @param geohash The geohash of the location.
+ * @returns The location encoded as GeoFire object.
+ */
+export function encodeGeoFireObject(location: number[], geohash: string): GeoFireObj {
+  validateLocation(location);
+  validateGeohash(geohash);
+  return { '.priority': geohash, 'g': geohash, 'l': location };
+}
+
+/**
+ * Decodes the location given as GeoFire object. Returns null if decoding fails.
+ *
+ * @param geoFireObj The location encoded as GeoFire object.
+ * @returns The location as [latitude, longitude] pair or null if decoding fails.
+ */
+export function decodeGeoFireObject(geoFireObj: GeoFireObj): number[] {
+  if (geoFireObj && 'l' in geoFireObj && Array.isArray(geoFireObj.l) && geoFireObj.l.length === 2) {
+    return geoFireObj.l;
+  } else {
+    throw new Error('Unexpected location object encountered: ' + JSON.stringify(geoFireObj));
+  }
+}
+
+/**
+ * Returns the key of a Firebase snapshot across SDK versions.
+ *
+ * @param A Firebase snapshot.
+ * @returns The Firebase snapshot's key.
+ */
+export function geoFireGetKey(snapshot: firebase.database.DataSnapshot): string {
+  let key: string;
```

Comment:
It looks like part of the function body was removed here (specifically the handling of ` snapshot.key` fxn and a fallback `snapshot.name`. 

What was the rationale?

Reply:
I didn't realize that those were supported keys/attributes in older version of Firebase. I was basing it off of the typing from Firebase 4.13.1. Anyway I'll be adding it back in.